Multi facet selection - Record last user state

rominail opened the pull request:
For the multi facets js selection, record user choice in local storage

EreMaijala commented on themes/bootstrap3/js/facets.js:
```diff
@@ -140,6 +140,7 @@ VuFind.register('multiFacetsSelection', function multiFacetsSelection() {
   const activation_event = 'facet-selection-begin';
   const deactivation_event = 'facet-selection-cancel';
   const apply_event = 'facet-selection-done';
+  const localStorageVariableName = 'multiFacetsSelection';
```
Other consts are in snake_case

EreMaijala commented on themes/bootstrap3/js/facets.js:
```diff
@@ -313,6 +318,7 @@ VuFind.register('multiFacetsSelection', function multiFacetsSelection() {
         toggleSelectedFacetStyle(elem);
       }
     }
+    saveUserSelectionLastState(isMultiFacetsSelectionActivated);
```
While it doesn't hurt here, I think the proper place for this is inside the `if (typeof enable !== 'undefined') {` block above.

EreMaijala commented on themes/bootstrap3/js/facets.js:
```diff
@@ -338,6 +344,7 @@ VuFind.register('multiFacetsSelection', function multiFacetsSelection() {
     const activationElem = context.querySelector('.js-user-selection-multi-filters');
     if (activationElem) {
       activationElem.addEventListener('change', function multiFacetSelectionChange() { toggleMultiFacetsSelection(this.checked); } );
+      toggleMultiFacetsSelection(localStorage.getItem(localStorageVariableName) === 'true');
```
Since there's `saveUserSelectionLastState`, there should also be `getUserSelectionLastState` or such to retrieve the selected state (as a boolean).

EreMaijala commented on themes/bootstrap3/js/facets.js:
```diff
@@ -297,6 +298,10 @@ VuFind.register('multiFacetsSelection', function multiFacetsSelection() {
+
   function toggleMultiFacetsSelection(enable) {
     if (typeof enable !== 'undefined') {
       isMultiFacetsSelectionActivated = enable;
```
I think we need to return early here if `enable === isMultiFacetsSelectionActivated`. Otherwise on initialization this function can run with `enable === false` without any previous activation and do useless work and emit an event that doesn't reflect reality.

rominail replied:
I don't understand what you want you mean

EreMaijala replied:
`initMultiFacetControls` calls `toggleMultiFacetsSelection` upon initialization. if `getUserSelectionLastState()` returns false, it causes `toggleMultiFacetsSelection` to act like multi-facet selection had been active, but it hadn't been. If you do this, it won't do anything when the requested state is already active:

```
  function toggleMultiFacetsSelection(enable) {
    if (typeof enable !== 'undefined') {
      if (isMultiFacetsSelectionActivated === enabled) {
        return;
      }
      isMultiFacetsSelectionActivated = enable;
```

rominail replied:
Thanks for the explanation; I fixed it

EreMaijala commented on themes/bootstrap3/js/facets.js:
```diff
@@ -297,9 +298,21 @@ VuFind.register('multiFacetsSelection', function multiFacetsSelection() {
   }
 
+  function saveUserSelectionLastState(state) {
+    localStorage.setItem(local_storage_variable_name, state);
```
Noticed one more thing: While this may work, I think it would be more appropriate to set the value explicitly to `state ? 'true' : 'false'`) since `setItem` takes a string but `state` is a boolean.

demiankatz reviewed (changes requested):
@EreMaijala, are tests passing for you? Mine are consistently failing with:

```
There were 2 failures:

1) VuFindTest\Mink\SearchFacetsTest::testCheckboxFacetSelection with data set "select one (multi/single)" (['Books'], 8, true, false)
Failed asserting that 'Showing 1 - 8 results of 8' [ASCII](length: 26) contains "Showing 1 - 19 results" [ASCII](length: 22).

/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php:1224

2) VuFindTest\Mink\SearchFacetsTest::testCheckboxFacetSelection with data set "select two (multi/single)" (['Books', 'Fiction'], 7, true, false)
Failed asserting that 'Showing 1 - 7 results of 7' [ASCII](length: 26) contains "Showing 1 - 19 results" [ASCII](length: 22).

/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/SearchFacetsTest.php:1224
```

I haven't taken a close look at why this might be -- not sure if it's a bug in the code or in the test, assuming it's not just a local glitch.

Please let me know if you need me to do more investigation on my end; I thought it would be helpful to get a second opinion before diving in too deep!

EreMaijala commented:
@demiankatz Oops, I failed to consider the tests. They will obviously need to be updated because they assume multi-selection is inactive on each page load.

demiankatz commented:
@rominail, are you interested in trying to figure this out, or do you want help? (I don't think you've worked with the test suite before, so I'm happy to help if you don't have time to dive in at the moment -- but if you're interested in learning, this might be a good opportunity to do so).

rominail commented:
I just realized I totally missed the updates here
I'll take a look into it and let you know if I need help

rominail committed:
Merge to last code

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into multiFacetsSelection

demiankatz commented:
@rominail, I noticed that there was a conflict here caused by the removal of the bootstrap3 theme; I went ahead and resolved it for you. It will be nice no longer having to edit two files for every change! Please let me know if you need anything else from my end....

rominail commented:
Thanks, I'm working on setting up a docker image of vufind locally in order to test my PR, I'll let you know when I'm done or if I have troubles

rominail committed:
Merge branch 'dev' into multiFacetsSelection

rominail committed:
Integration tests

rominail commented:
I added a test and modified the other ones
I changed others which are not related, because I realized when you mess around with all the tests, i.e. exit in the middle of one, if you re run the tests  the config will be different and it will raise bugs or tests fail
Let me know what you think of it

demiankatz commented:
Thanks, @rominail, I'll take a closer look at this soon!

For future reference, there is a Phing task called `reset_setup` which will restore the test environment's database and configuration settings to a default state. This can be helpful to get things back to normal after a test is interrupted in an unexpected way.

demiankatz committed:
Merge branch 'dev' into multiFacetsSelection

demiankatz committed:
Improve types.

demiankatz committed:
Improve assertions.

demiankatz reviewed (approved):
Thanks, @rominail, this is looking good! I just pushed up a couple of small commits -- one to make the types in your new trait methods slightly more specific, and one to improve the assertions in testMultiFacetsSelectionPersistence (I'm not sure why you were using assertStringContainsString to test a Boolean value, so I simplified that and added checks both before and after toggling the control to ensure we're checking the state in both situations).

I'm running the full test suite now and I expect it to pass. Unless something goes wrong and surprises me, I'll push the merge button on this within the next few minutes.
